system/jlib/jptree.cpp

```diff
@@ -7442,6 +7442,7 @@ class CJSONReader : public CJSONReaderBase<X>, implements IPTreeReader
                 {
                     if ('\0' == *(name+6)) // #value
                     {
+                        assertex(retValue);
```

Comment:
Again, this feels like the warning is highlighting a bug, and adding code to make it go away is a bad idea. The bug in this case being (as far as I can tell from a very quick glance at the code) that the error checking when parsing malformed JSON is inadequate.

Reply:
this was the other one Russ spoke about:
```
Consolidate compiler generated dependencies of target jlib
[  0%] Building CXX object system/jlib/CMakeFiles/jlib.dir/jptree.cpp.o
In member function ‘void CJSONReader<X>::readValueNotify(const char*, bool, StringBuffer*, bool*) [with X = CInstBufferReader]’,
inlined from ‘void CJSONReader<X>::readArray(const char*) [with X = CInstBufferReader]’ at /mnt/e/HPCC/HPCC-Platform/system/jlib/jptree.cpp:7481:32:
/mnt/e/HPCC/HPCC-Platform/system/jlib/jptree.cpp:7445:43: error: ‘this’ pointer is null [-Werror=nonnull]
7445 |                         retValue->swapWith(value);
|                         ~~~~~~^
In file included from /mnt/e/HPCC/HPCC-Platform/system/jlib/jlib.hpp:276,
from /mnt/e/HPCC/HPCC-Platform/system/jlib/jstats.h:22,
```

There is already a dbgassertex(retValue && isValueBinary); a few lines above, but _DEBUG only.
This function is called with retValue is null.

I suspect it's a code path that's never expect and the original assert is just a guard against the unexpected, but @afishbeck may know more.

Reply:
There is a call that passes null for this parameter. Unless you can be sure that the json being parsed at the point that call is made CANNOT have #value in it , then this warning is highlighting inadequate error checking, and suppressing the warning is wrong.

Reply:
First a little background.  This #value is meant for specific ptree serialization scenarios, not as a syntax for users.  Where we do XML -> ptree -> JSON -> ptree. 

XML supports "mixed content" which is formatted like this:

`<a>extra content1<b>1</b>extra content2</a>`
Notice `<a>` has both a child and content of it's own.

Inside ptree those bits of extra content are merged together, so it turns into:

`<a><b>1</b>extra content1extra content2</a>`

JSON doesn't support any extra content.  There is nowhere to put it:

`{"a": {"b": 1}}`

So to preserve all the ptree data in different forms I added the #value concept... 

`{"a": {"b": 1, "#value": "extra content1extra content2"}}`

It's only meant to support what ptree would have written out.  Of course someone could try to hand code these values into their JSON.

The scenarios where retValue could be null here would likely to have been typed or generated some other way:

An array of #value (PTREE would have only output one entry:

`{"a": {"b": 1, "#value": ["extra", "content1", "extra", "content2"], "c": 2 }}
`
Or a rootless xml document that has #value at the root.

Reply:
Bottom line, I think we can just add a check before using retValue.

`
if (retValue)
  retValue->swapWith(value);
`
Even if we wanted to support 
`{"a": {"b": 1, "#value": ["extra", "content1", "extra", "content2"], "c": 2 }}`
which shouldn't be output by toJSON(), it would be outside the scope of this PR.